Is there a way to run doctests and continue on failure?

Say you have a file named t.py:

``` python
"""
>>> print(1)
>>> print(1)
>>> print(1)
"""

if __name__ == "__main__":
    import doctest
    doctest.testmod()
```

When I run this, I would get:

```
$ python t.py
**********************************************************************
File "t.py", line 2, in __main__
Failed example:
    print(1)
Expected nothing
Got:
    1
**********************************************************************
File "t.py", line 3, in __main__
Failed example:
    print(1)
Expected nothing
Got:
    1
**********************************************************************
File "t.py", line 4, in __main__
Failed example:
    print(1)
Expected nothing
Got:
    1
**********************************************************************
1 items had failures:
   3 of   3 in __main__
***Test Failed*** 3 failures.
```
However, when I run with pytest, I would only get the first failure:

```
$ pytest t.py
========================================================= test session starts =========================================================
platform linux2 -- Python 2.7.14, pytest-3.3.2, py-1.5.2, pluggy-0.5.2
rootdir: /auto/cnvtvws/wlee/fcnvt/dates, inifile: setup.cfg
plugins: xdist-1.20.1, forked-0.2, flakes-1.0.1, flake8-0.9, cov-2.5.1, hypothesis-3.38.5, pep8-1.0.6
collected 2 items

t.py sF                                                                                                                         [100%]
======================================================= short test summary info =======================================================
FAIL t.py::t

============================================================== FAILURES ===============================================================
_____________________________________________________________ [doctest] t _____________________________________________________________
001
002 >>> print(1)
Expected nothing
Got:
    1

.../t.py:2: DocTestFailure
================================================= 1 failed, 1 skipped in 0.01 seconds =================================================
```

I've tried to dig through all the options form the documentation but I wan't able to figure it out.  Is there a way to have pytest report all the failures at once?  Otherwise I'd have to fix the failures one by one, which gets tedious.  I tried to put in various entries in --doctest-report='none' but can't get it to work.

Here is the relevant section in setup.cfg:

```
[tool:pytest]
addopts = -rf --strict --doctest-modules --flake8 --doctest-glob='*.rst'

doctest_optionflags = NORMALIZE_WHITESPACE
maxfail = 5
```
Here is my pip list:

```
alabaster (0.7.10)
asn1crypto (0.24.0)
attrs (17.3.0)
Babel (2.5.3)
certifi (2017.11.5)
cffi (1.11.4)
chardet (3.0.4)
configparser (3.5.0)
coverage (4.4.2)
cryptography (2.1.4)
docutils (0.14)
enum34 (1.1.6)
execnet (1.2.0)
flake8 (3.5.0)
funcsigs (1.0.2)
hypothesis (3.38.5)
idna (2.6)
imagesize (0.7.1)
ipaddress (1.0.19)
Jinja2 (2.10)
llvmlite (0.15.0)
MarkupSafe (1.0)
mccabe (0.6.1)
numba (0.30.1+0.g8c1033f.dirty)
numpy (1.11.3)
pep8 (1.7.1)
pip (9.0.1)
pluggy (0.6.0)
py (1.5.2)
pycodestyle (2.3.1)
pycparser (2.18)
pyflakes (1.6.0)
Pygments (2.2.0)
Pympler (0.5)
pyOpenSSL (17.5.0)
PySocks (1.6.7)
pytest (3.3.2)
pytest-cache (1.0)
pytest-cov (2.5.1)
pytest-flake8 (0.9)
pytest-flakes (1.0.1)
pytest-forked (0.2)
pytest-pep8 (1.0.6)
pytest-runner (3.0)
pytest-xdist (1.20.1)
python-dateutil (2.4.1)
pytz (2015.7)
requests (2.18.4)
setuptools (38.4.0)
singledispatch (3.4.0.3)
six (1.11.0)
snowballstemmer (1.2.1)
Sphinx (1.6.6)
sphinxcontrib-websupport (1.0.1)
typing (3.6.2)
urllib3 (1.22)
wheel (0.30.0)
zope.interface (4.4.3)
```

--- a/AUTHORS
+++ b/AUTHORS
@@ -195,6 +195,7 @@ Victor Uriarte
 Vidar T. Fauske
 Vitaly Lashmanov
 Vlad Dragos
+William Lee
 Wouter van Ackooy
 Xuan Luong
 Xuecong Liao

--- a/_pytest/doctest.py
+++ b/_pytest/doctest.py
@@ -24,6 +24,9 @@ DOCTEST_REPORT_CHOICES = (
     DOCTEST_REPORT_CHOICE_ONLY_FIRST_FAILURE,
 )
 
+# Lazy definiton of runner class
+RUNNER_CLASS = None
+
 
 def pytest_addoption(parser):
     parser.addini('doctest_optionflags', 'option flags for doctests',
@@ -77,14 +80,91 @@ def _is_doctest(config, path, parent):
 
 class ReprFailDoctest(TerminalRepr):
 
-    def __init__(self, reprlocation, lines):
-        self.reprlocation = reprlocation
-        self.lines = lines
+    def __init__(self, reprlocation_lines):
+        # List of (reprlocation, lines) tuples
+        self.reprlocation_lines = reprlocation_lines
 
     def toterminal(self, tw):
-        for line in self.lines:
-            tw.line(line)
-        self.reprlocation.toterminal(tw)
+        for reprlocation, lines in self.reprlocation_lines:
+            for line in lines:
+                tw.line(line)
+            reprlocation.toterminal(tw)
+
+
+# class DoctestFailureContainer(object):
+#
+#     NAME = 'DocTestFailure'
+#
+#     def __init__(self, test, example, got):
+#         self.test = test
+#         self.example = example
+#         self.got = got
+#
+#
+# class DoctestUnexpectedExceptionContainer(object):
+#
+#     NAME = 'DoctestUnexpectedException'
+#
+#     def __init__(self, test, example, exc_info):
+#         self.test = test
+#         self.example = example
+#         self.exc_info = exc_info
+
+
+class MultipleDoctestFailures(Exception):
+    def __init__(self, failures):
+        super(MultipleDoctestFailures, self).__init__()
+        self.failures = failures
+
+
+def _init_runner_class():
+    import doctest
+
+    class PytestDoctestRunner(doctest.DocTestRunner):
+        """
+        Runner to collect failures.  Note that the out variable in this case is
+        a list instead of a stdout-like object
+        """
+        def __init__(self, checker=None, verbose=None, optionflags=0,
+                     continue_on_failure=True):
+            doctest.DocTestRunner.__init__(
+                self, checker=checker, verbose=verbose, optionflags=optionflags)
+            self.continue_on_failure = continue_on_failure
+
+        def report_start(self, out, test, example):
+            pass
+
+        def report_success(self, out, test, example, got):
+            pass
+
+        def report_failure(self, out, test, example, got):
+            # failure = DoctestFailureContainer(test, example, got)
+            failure = doctest.DocTestFailure(test, example, got)
+            if self.continue_on_failure:
+                out.append(failure)
+            else:
+                raise failure
+
+        def report_unexpected_exception(self, out, test, example, exc_info):
+            # failure = DoctestUnexpectedExceptionContainer(test, example, exc_info)
+            failure = doctest.UnexpectedException(test, example, exc_info)
+            if self.continue_on_failure:
+                out.append(failure)
+            else:
+                raise failure
+
+    return PytestDoctestRunner
+
+
+def _get_runner(checker=None, verbose=None, optionflags=0,
+                continue_on_failure=True):
+    # We need this in order to do a lazy import on doctest
+    global RUNNER_CLASS
+    if RUNNER_CLASS is None:
+        RUNNER_CLASS = _init_runner_class()
+    return RUNNER_CLASS(
+        checker=checker, verbose=verbose, optionflags=optionflags,
+        continue_on_failure=continue_on_failure)
 
 
 class DoctestItem(pytest.Item):
@@ -106,7 +186,10 @@ class DoctestItem(pytest.Item):
     def runtest(self):
         _check_all_skipped(self.dtest)
         self._disable_output_capturing_for_darwin()
-        self.runner.run(self.dtest)
+        failures = []
+        self.runner.run(self.dtest, out=failures)
+        if failures:
+            raise MultipleDoctestFailures(failures)
 
     def _disable_output_capturing_for_darwin(self):
         """
@@ -122,42 +205,51 @@ class DoctestItem(pytest.Item):
 
     def repr_failure(self, excinfo):
         import doctest
+        failures = None
         if excinfo.errisinstance((doctest.DocTestFailure,
                                   doctest.UnexpectedException)):
-            doctestfailure = excinfo.value
-            example = doctestfailure.example
-            test = doctestfailure.test
-            filename = test.filename
-            if test.lineno is None:
-                lineno = None
-            else:
-                lineno = test.lineno + example.lineno + 1
-            message = excinfo.type.__name__
-            reprlocation = ReprFileLocation(filename, lineno, message)
-            checker = _get_checker()
-            report_choice = _get_report_choice(self.config.getoption("doctestreport"))
-            if lineno is not None:
-                lines = doctestfailure.test.docstring.splitlines(False)
-                # add line numbers to the left of the error message
-                lines = ["%03d %s" % (i + test.lineno + 1, x)
-                         for (i, x) in enumerate(lines)]
-                # trim docstring error lines to 10
-                lines = lines[max(example.lineno - 9, 0):example.lineno + 1]
-            else:
-                lines = ['EXAMPLE LOCATION UNKNOWN, not showing all tests of that example']
-                indent = '>>>'
-                for line in example.source.splitlines():
-                    lines.append('??? %s %s' % (indent, line))
-                    indent = '...'
-            if excinfo.errisinstance(doctest.DocTestFailure):
-                lines += checker.output_difference(example,
-                                                   doctestfailure.got, report_choice).split("\n")
-            else:
-                inner_excinfo = ExceptionInfo(excinfo.value.exc_info)
-                lines += ["UNEXPECTED EXCEPTION: %s" %
-                          repr(inner_excinfo.value)]
-                lines += traceback.format_exception(*excinfo.value.exc_info)
-            return ReprFailDoctest(reprlocation, lines)
+            failures = [excinfo.value]
+        elif excinfo.errisinstance(MultipleDoctestFailures):
+            failures = excinfo.value.failures
+
+        if failures is not None:
+            reprlocation_lines = []
+            for failure in failures:
+                example = failure.example
+                test = failure.test
+                filename = test.filename
+                if test.lineno is None:
+                    lineno = None
+                else:
+                    lineno = test.lineno + example.lineno + 1
+                message = type(failure).__name__
+                reprlocation = ReprFileLocation(filename, lineno, message)
+                checker = _get_checker()
+                report_choice = _get_report_choice(self.config.getoption("doctestreport"))
+                if lineno is not None:
+                    lines = failure.test.docstring.splitlines(False)
+                    # add line numbers to the left of the error message
+                    lines = ["%03d %s" % (i + test.lineno + 1, x)
+                             for (i, x) in enumerate(lines)]
+                    # trim docstring error lines to 10
+                    lines = lines[max(example.lineno - 9, 0):example.lineno + 1]
+                else:
+                    lines = ['EXAMPLE LOCATION UNKNOWN, not showing all tests of that example']
+                    indent = '>>>'
+                    for line in example.source.splitlines():
+                        lines.append('??? %s %s' % (indent, line))
+                        indent = '...'
+                if isinstance(failure, doctest.DocTestFailure):
+                    lines += checker.output_difference(example,
+                                                       failure.got,
+                                                       report_choice).split("\n")
+                else:
+                    inner_excinfo = ExceptionInfo(failure.exc_info)
+                    lines += ["UNEXPECTED EXCEPTION: %s" %
+                              repr(inner_excinfo.value)]
+                    lines += traceback.format_exception(*failure.exc_info)
+                reprlocation_lines.append((reprlocation, lines))
+            return ReprFailDoctest(reprlocation_lines)
         else:
             return super(DoctestItem, self).repr_failure(excinfo)
 
@@ -202,8 +294,10 @@ class DoctestTextfile(pytest.Module):
         globs = {'__name__': '__main__'}
 
         optionflags = get_optionflags(self)
-        runner = doctest.DebugRunner(verbose=0, optionflags=optionflags,
-                                     checker=_get_checker())
+        continue_on_failure = not self.config.getvalue("usepdb")
+        runner = _get_runner(verbose=0, optionflags=optionflags,
+                             checker=_get_checker(),
+                             continue_on_failure=continue_on_failure)
         _fix_spoof_python2(runner, encoding)
 
         parser = doctest.DocTestParser()
@@ -238,8 +332,10 @@ class DoctestModule(pytest.Module):
         # uses internal doctest module parsing mechanism
         finder = doctest.DocTestFinder()
         optionflags = get_optionflags(self)
-        runner = doctest.DebugRunner(verbose=0, optionflags=optionflags,
-                                     checker=_get_checker())
+        continue_on_failure = not self.config.getvalue("usepdb")
+        runner = _get_runner(verbose=0, optionflags=optionflags,
+                             checker=_get_checker(),
+                             continue_on_failure=continue_on_failure)
 
         for test in finder.find(module, module.__name__):
             if test.examples:  # skip empty doctests

--- a/_pytest/doctest.py
+++ b/_pytest/doctest.py
@@ -91,26 +91,6 @@ class ReprFailDoctest(TerminalRepr):
             reprlocation.toterminal(tw)
 
 
-# class DoctestFailureContainer(object):
-#
-#     NAME = 'DocTestFailure'
-#
-#     def __init__(self, test, example, got):
-#         self.test = test
-#         self.example = example
-#         self.got = got
-#
-#
-# class DoctestUnexpectedExceptionContainer(object):
-#
-#     NAME = 'DoctestUnexpectedException'
-#
-#     def __init__(self, test, example, exc_info):
-#         self.test = test
-#         self.example = example
-#         self.exc_info = exc_info
-
-
 class MultipleDoctestFailures(Exception):
     def __init__(self, failures):
         super(MultipleDoctestFailures, self).__init__()
@@ -138,7 +118,6 @@ def _init_runner_class():
             pass
 
         def report_failure(self, out, test, example, got):
-            # failure = DoctestFailureContainer(test, example, got)
             failure = doctest.DocTestFailure(test, example, got)
             if self.continue_on_failure:
                 out.append(failure)
@@ -146,7 +125,6 @@ def _init_runner_class():
                 raise failure
 
         def report_unexpected_exception(self, out, test, example, exc_info):
-            # failure = DoctestUnexpectedExceptionContainer(test, example, exc_info)
             failure = doctest.UnexpectedException(test, example, exc_info)
             if self.continue_on_failure:
                 out.append(failure)

--- a/_pytest/doctest.py
+++ b/_pytest/doctest.py
@@ -50,6 +50,10 @@ def pytest_addoption(parser):
                     action="store_true", default=False,
                     help="ignore doctest ImportErrors",
                     dest="doctest_ignore_import_errors")
+    group.addoption("--doctest-continue-on-failure",
+                    action="store_true", default=False,
+                    help="for a given doctest, continue to run after the first failure",
+                    dest="doctest_continue_on_failure")
 
 
 def pytest_collect_file(path, parent):
@@ -100,22 +104,16 @@ class MultipleDoctestFailures(Exception):
 def _init_runner_class():
     import doctest
 
-    class PytestDoctestRunner(doctest.DocTestRunner):
+    class PytestDoctestRunner(doctest.DebugRunner):
         """
         Runner to collect failures.  Note that the out variable in this case is
         a list instead of a stdout-like object
         """
         def __init__(self, checker=None, verbose=None, optionflags=0,
                      continue_on_failure=True):
-            doctest.DocTestRunner.__init__(
+            doctest.DebugRunner.__init__(
                 self, checker=checker, verbose=verbose, optionflags=optionflags)
             self.continue_on_failure = continue_on_failure
-
-        def report_start(self, out, test, example):
-            pass
-
-        def report_success(self, out, test, example, got):
-            pass
 
         def report_failure(self, out, test, example, got):
             failure = doctest.DocTestFailure(test, example, got)
@@ -257,6 +255,16 @@ def get_optionflags(parent):
     return flag_acc
 
 
+def _get_continue_on_failure(config):
+    continue_on_failure = config.getvalue('doctest_continue_on_failure')
+    if continue_on_failure:
+        # We need to turn off this if we use pdb since we should stop at
+        # the first failure
+        if config.getvalue("usepdb"):
+            continue_on_failure = False
+    return continue_on_failure
+
+
 class DoctestTextfile(pytest.Module):
     obj = None
 
@@ -272,10 +280,11 @@ class DoctestTextfile(pytest.Module):
         globs = {'__name__': '__main__'}
 
         optionflags = get_optionflags(self)
-        continue_on_failure = not self.config.getvalue("usepdb")
-        runner = _get_runner(verbose=0, optionflags=optionflags,
-                             checker=_get_checker(),
-                             continue_on_failure=continue_on_failure)
+
+        runner = _get_runner(
+            verbose=0, optionflags=optionflags,
+            checker=_get_checker(),
+            continue_on_failure=_get_continue_on_failure(self.config))
         _fix_spoof_python2(runner, encoding)
 
         parser = doctest.DocTestParser()
@@ -310,10 +319,10 @@ class DoctestModule(pytest.Module):
         # uses internal doctest module parsing mechanism
         finder = doctest.DocTestFinder()
         optionflags = get_optionflags(self)
-        continue_on_failure = not self.config.getvalue("usepdb")
-        runner = _get_runner(verbose=0, optionflags=optionflags,
-                             checker=_get_checker(),
-                             continue_on_failure=continue_on_failure)
+        runner = _get_runner(
+            verbose=0, optionflags=optionflags,
+            checker=_get_checker(),
+            continue_on_failure=_get_continue_on_failure(self.config))
 
         for test in finder.find(module, module.__name__):
             if test.examples:  # skip empty doctests

--- a/changelog/3149.feature
+++ b/changelog/3149.feature
@@ -1,0 +1,1 @@
+Doctest runs now show multiple failures for each doctest snippet, instead of stopping at the first failure.

--- a/changelog/3149.feature
+++ b/changelog/3149.feature
@@ -1,1 +1,1 @@
-Doctest runs now show multiple failures for each doctest snippet, instead of stopping at the first failure.
+New ``--doctest-continue-on-failure`` command-line option to enable doctests to show multiple failures for each snippet, instead of stopping at the first failure.

--- a/doc/en/doctest.rst
+++ b/doc/en/doctest.rst
@@ -115,6 +115,11 @@ itself::
     >>> get_unicode_greeting()  # doctest: +ALLOW_UNICODE
     'Hello'
 
+By default, pytest would report only the first failure for a given doctest.  If
+you want to continue the test even when you have failures, do::
+
+    pytest --doctest-modules --doctest-continue-on-failure
+
 
 The 'doctest_namespace' fixture
 -------------------------------

--- a/testing/test_doctest.py
+++ b/testing/test_doctest.py
@@ -769,11 +769,13 @@ class TestDoctestSkips(object):
         """)
         result = testdir.runpytest("--doctest-modules")
         result.assert_outcomes(passed=0, failed=1)
-        # We need to make sure we have two failure lines (4, 5, and 8) instead of
-        # one.
-        result.stdout.fnmatch_lines("*test_something.txt:4: DoctestUnexpectedException*")
-        result.stdout.fnmatch_lines("*test_something.txt:5: DocTestFailure*")
-        result.stdout.fnmatch_lines("*test_something.txt:8: DocTestFailure*")
+        # The lines that contains the failure are 4, 5, and 8.  The first one
+        # is a stack trace and the other two are mismatches.
+        result.stdout.fnmatch_lines([
+            "*4: UnexpectedException*",
+            "*5: DocTestFailure*",
+            "*8: DocTestFailure*",
+        ])
 
 
 class TestDoctestAutoUseFixtures(object):

--- a/testing/test_doctest.py
+++ b/testing/test_doctest.py
@@ -756,6 +756,25 @@ class TestDoctestSkips(object):
         reprec = testdir.inline_run("--doctest-modules")
         reprec.assertoutcome(passed=0, skipped=0)
 
+    def test_continue_on_failure(self, testdir):
+        testdir.maketxtfile(test_something="""
+            >>> i = 5
+            >>> def foo():
+            ...     raise ValueError('error1')
+            >>> foo()
+            >>> i
+            >>> i + 2
+            7
+            >>> i + 1
+        """)
+        result = testdir.runpytest("--doctest-modules")
+        result.assert_outcomes(passed=0, failed=1)
+        # We need to make sure we have two failure lines (4, 5, and 8) instead of
+        # one.
+        result.stdout.fnmatch_lines("*test_something.txt:4: DoctestUnexpectedException*")
+        result.stdout.fnmatch_lines("*test_something.txt:5: DocTestFailure*")
+        result.stdout.fnmatch_lines("*test_something.txt:8: DocTestFailure*")
+
 
 class TestDoctestAutoUseFixtures(object):
 

--- a/testing/test_doctest.py
+++ b/testing/test_doctest.py
@@ -767,7 +767,7 @@ class TestDoctestSkips(object):
             7
             >>> i + 1
         """)
-        result = testdir.runpytest("--doctest-modules")
+        result = testdir.runpytest("--doctest-modules", "--doctest-continue-on-failure")
         result.assert_outcomes(passed=0, failed=1)
         # The lines that contains the failure are 4, 5, and 8.  The first one
         # is a stack trace and the other two are mismatches.

